test/lit/passes/type-ssa-exact-rmw.wast

```diff
@@ -0,0 +1,179 @@
+    ;; CHECK:       (type $expected-in-struct-cmpxchg (sub (struct (field anyref))))
+    (type $expected-in-struct-cmpxchg (sub (struct (field anyref))))
+    ;; CHECK:       (type $used-in-struct-cmpxchg (sub $expected-in-struct-cmpxchg (struct (field anyref))))
+    (type $used-in-struct-cmpxchg (sub $expected-in-struct-cmpxchg (struct (field anyref))))
+    ;; CHECK:       (type $used-in-struct-cmpxchg-ok (sub $expected-in-struct-cmpxchg (struct (field anyref))))
+    (type $used-in-struct-cmpxchg-ok (sub $expected-in-struct-cmpxchg (struct (field anyref))))
+  )
+
+  ;; CHECK:      (type $struct-rmw-exact (struct (field (mut (ref (exact $used-in-struct-rmw))))))
+  (type $struct-rmw-exact (struct (field (mut (ref (exact $used-in-struct-rmw))))))
+  ;; CHECK:      (type $struct-rmw-inexact (struct (field (mut (ref $used-in-struct-rmw-ok)))))
+  (type $struct-rmw-inexact (struct (field (mut (ref $used-in-struct-rmw-ok)))))
+
+  ;; CHECK:      (type $struct-cmpxchg-exact (struct (field (mut (ref (exact $used-in-struct-cmpxchg))))))
+  (type $struct-cmpxchg-exact (struct (field (mut (ref (exact $used-in-struct-cmpxchg))))))
+  ;; CHECK:      (type $struct-cmpxchg-inexact (struct (field (mut (ref $used-in-struct-cmpxchg-ok)))))
+  (type $struct-cmpxchg-inexact (struct (field (mut (ref $used-in-struct-cmpxchg-ok)))))
+
+  ;; CHECK:      (type $9 (func (param (ref (exact $used-in-struct-rmw)) (ref $struct-rmw-exact))))
+
+  ;; CHECK:      (type $10 (func (param (ref (exact $used-in-struct-rmw-ok)) (ref $struct-rmw-inexact))))
+
+  ;; CHECK:      (type $11 (func (param (ref (exact $used-in-struct-cmpxchg)) (ref (exact $expected-in-struct-cmpxchg)) (ref $struct-cmpxchg-exact))))
+
+  ;; CHECK:      (type $12 (func (param (ref (exact $used-in-struct-cmpxchg-ok)) (ref (exact $expected-in-struct-cmpxchg)) (ref $struct-cmpxchg-inexact))))
+
+  ;; CHECK:      (rec
+  ;; CHECK-NEXT:  (type $used-in-struct-rmw-ok_1 (sub $used-in-struct-rmw-ok (struct (field anyref))))
+
+  ;; CHECK:       (type $expected-in-struct-cmpxchg_2 (sub $expected-in-struct-cmpxchg (struct (field anyref))))
+
+  ;; CHECK:       (type $expected-in-struct-cmpxchg_3 (sub $expected-in-struct-cmpxchg (struct (field anyref))))
+
+  ;; CHECK:       (type $used-in-struct-cmpxchg-ok_4 (sub $used-in-struct-cmpxchg-ok (struct (field anyref))))
+
+  ;; CHECK:      (func $struct-rmw (type $9) (param $used (ref (exact $used-in-struct-rmw))) (param $ref (ref $struct-rmw-exact))
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.atomic.rmw.xchg $struct-rmw-exact 0
+  ;; CHECK-NEXT:    (local.get $ref)
+  ;; CHECK-NEXT:    (local.get $used)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.new $used-in-struct-rmw
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $struct-rmw (param $used (ref (exact $used-in-struct-rmw))) (param $ref (ref $struct-rmw-exact))
+    (drop
```

Comment:
Please add comments in these tests. Reading this first one, I'm not sure how the parameters are relevant, or how the two instructions relate to each other.

(I probably didn't fully understand the problem when you described it before - I was thinking the issue is something like the output of a `struct.new` used with exact type, but the `struct.new` here is dropped.)

Reply:
To determine whether individual `struct.news` can be optimized or not would require a data flow analysis to prove that a particular allocation never flows to a location that observes its exactness. We don't do any such analysis, so the best we can do is to inhibit optimization for allocations of any type that is ever observed to be exact, regardless of the data flow from those allocations.

Reply:
I see, thanks. Makes sense now.